File: li2/li2/plp/imperative2/exemplos/Exemplo1.java
package li2.plp.imperative2.exemplos;

import li2.plp.expressions2.expression.*;
import li2.plp.imperative1.command.*;
import li2.plp.imperative1.declaration.*;
import li2.plp.imperative1.memory.ListaValor;
import li2.plp.imperative2.Programa;
import li2.plp.imperative2.command.*;
import li2.plp.imperative2.declaration.*;
import li2.plp.imperative2.memory.*;

/**
 * Atividade 1.
 *
 * Constroi e executa, usando a arvore sintatica abstrata da li2, o comando:
 *
 *   { var a = 0, proc incA () {a := a + 1};
 *     call incA(); call incA(); write(a) }
 *
 * Saida esperada: 2 (consequencia das duas chamadas ao procedimento).
 */
public class Exemplo1 {

	public static void main(String[] args) throws Exception {

		// ---------- Declaracoes do bloco ----------
		// var a = 0
		Declaracao declaracaoVariavel =
				new DeclaracaoVariavel(new Id("a"), new ValorInteiro(0));

		// proc incA () { a := a + 1 }
		DefProcedimento defIncA = new DefProcedimento(
				new ListaDeclaracaoParametro(),                 // sem parametros
				new Atribuicao(new Id("a"),
						new ExpSoma(new Id("a"), new ValorInteiro(1))));
		Declaracao declaracaoProcedimento =
				new DeclaracaoProcedimento(new Id("incA"), defIncA);

		// var a = 0, proc incA () {...}
		Declaracao declaracoes =
				new DeclaracaoComposta(declaracaoVariavel, declaracaoProcedimento);

		// ---------- Corpo do bloco ----------
		// call incA(); call incA(); write(a)
		Comando corpo = new SequenciaComando(
				new ChamadaProcedimento(new Id("incA"), new ListaExpressao()),
				new SequenciaComando(
						new ChamadaProcedimento(new Id("incA"), new ListaExpressao()),
						new Write(new Id("a"))));

		// ---------- Bloco completo ----------
		Comando programa = new ComandoDeclaracao(declaracoes, corpo);

		AmbienteExecucaoImperativa2 ambiente =
				new ContextoExecucaoImperativa2(new ListaValor());

		ListaValor saida = new Programa(programa).executar(ambiente);
		System.out.println("Exemplo1 -> saida: " + saida);  // esperado: 2
	}
}

File: python/li2/plp/imperative2/Exemplo1.py
"""
Atividade 1 (Lab 05 - li2).

Executa:
  { var a = 0, proc incA () {a := a + 1};
    call incA(); call incA(); write(a)
  }
"""
from li2.plp.expressions2.expression.ExpSoma import ExpSoma
from li2.plp.expressions2.expression.Id import Id
from li2.plp.expressions2.expression.ValorInteiro import ValorInteiro
from li2.plp.imperative1.command.Atribuicao import Atribuicao
from li2.plp.imperative1.command.ComandoDeclaracao import ComandoDeclaracao
from li2.plp.imperative1.command.SequenciaComando import SequenciaComando
from li2.plp.imperative1.command.Write import Write
from li2.plp.imperative1.declaration.DeclaracaoComposta import DeclaracaoComposta
from li2.plp.imperative1.declaration.DeclaracaoVariavel import DeclaracaoVariavel
from li2.plp.imperative1.memory.ContextoCompilacaoImperativa import ContextoCompilacaoImperativa
from li2.plp.imperative1.memory.ListaValor import ListaValor
from li2.plp.imperative2.command.ChamadaProcedimento import ChamadaProcedimento
from li2.plp.imperative2.command.ListaExpressao import ListaExpressao
from li2.plp.imperative2.declaration.DeclaracaoProcedimento import DeclaracaoProcedimento
from li2.plp.imperative2.declaration.DefProcedimento import DefProcedimento
from li2.plp.imperative2.declaration.ListaDeclaracaoParametro import ListaDeclaracaoParametro
from li2.plp.imperative2.memory.ContextoExecucaoImperativa2 import ContextoExecucaoImperativa2
from li2.plp.imperative2.Programa import Programa


class Exemplo1:

    @staticmethod
    def main():
        a = Id("a")
        inc_a = Id("incA")

        def_inc_a = DefProcedimento(ListaDeclaracaoParametro(), Atribuicao(a, ExpSoma(a, ValorInteiro(1))))
        declaracoes = DeclaracaoComposta(
            DeclaracaoVariavel(a, ValorInteiro(0)),
            DeclaracaoProcedimento(inc_a, def_inc_a),
        )
        chamadas = SequenciaComando(
            ChamadaProcedimento(inc_a, ListaExpressao()),
            SequenciaComando(ChamadaProcedimento(inc_a, ListaExpressao()), Write(a)),
        )
        comando = ComandoDeclaracao(declaracoes, chamadas)
        programa = Programa(comando)

        if programa.checaTipo(ContextoCompilacaoImperativa(ListaValor())):
            saida = programa.executar(ContextoExecucaoImperativa2(ListaValor()))
            print(saida)
        else:
            print("Comando mal tipado.")


if __name__ == "__main__":
    Exemplo1.main()
